Predict SbatLevelRT

lcp opened the pull request:
The 'SbatLevelRT' variable may change after shim update, and that changes PCR7. This PR looks into shim.efi and tries to predict 'SbatLevelRT' for the next boot so that the policy digest in the sealed key can be updated correctly.

lcp committed:
Locate .sbatlevel section in shim.efi

The .sbatlevel section in shim.efi will be used to set SbatLevelRT and
we need the section to predict SbatLevelRT.

Per PE format SPEC(*), the section name is

"An 8-byte, null-padded UTF-8 encoded string. If the string is exactly 8
 8 characters long, there is no terminating null. For longer names, this
 field contains a slash (/) that is followed by an ASCII representation
 of a decimal number that is an offset into the string table."

Before looking for the .sbatlevel section, we have to get the string
table, To get the offset to the string table:

  string_table_offset = symbol_table_offset +
                        (number_of_symbols * symbol_size)

With the offset, we can go further to look for the longer section names
and get the .sbatlevel section.

(*) https://learn.microsoft.com/en-us/windows/win32/debug/pe-format

Signed-off-by: Gary Lin <glin@suse.com>

lcp committed:
Predict SbatLevelRT for the next boot

The rehash ctx records the .sbatlevel section in the first EFI images.
When encountering 'SbatLevelRT' variable, sbat_automatic and sbat_latest
are extracted from the recorded .sbatlevel section for the candidate
SBAT level. The candidate SBAT level is sbat_automatic by default or
one of sbat_automatic, sbat_latest, SBAT_ORIGINAL based on 'SbatPolicy'
variable. If the datestamp of 'SbatLevelRT' is not later than that of
the candidate SBAT level, the candidate SBAT level will be used in the
next boot. Otherwise, just use the current SBAT level.

Signed-off-by: Gary Lin <glin@suse.com>

aplanas commented:
I think somehow this is breaking the PCR7 predictions on systems without secure boot (and pcr 7 is included in the policy)

lcp commented:
Could you paste the SbatLevelRT event from the following command?

`pcr-oracle -d -d --from eventlog --tpm-eventlog /sys/kernel/security/tpm0/binary_bios_measurements predict 7`

Maybe something went wrong with the SbatLevelRT event.

aplanas commented:
```
# pcr-oracle -d -d --from eventlog --tpm-eventlog /sys/kernel/security/tpm0/binary_bios_measurements predict 7
::: Initializing predictor for sha256:7 from eventlog
::: Detected TPMv2 event log
::: Successfully read 40 events from TPM event log
::: Created new predictor
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: runtime_read_efi_application(/dev/vda2, /EFI/systemd/shim.efi)
::: Read 965528 bytes from /tmp/efimnt.tIk2zn//EFI/systemd/shim.efi
::: Reading EFI application (/dev/vda2)/EFI/systemd/shim.efi
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: runtime_read_efi_application(/dev/vda2, /EFI/systemd/grub.efi)
::: Read 125808 bytes from /tmp/efimnt.LZI9xH//EFI/systemd/grub.efi
::: Reading EFI application (/dev/vda2)/EFI/systemd/grub.efi
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: runtime_read_efi_application(/dev/vda2, /opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6)
::: Read 15485296 bytes from /tmp/efimnt.XH06rJ//opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6
::: Reading EFI application (/dev/vda2)/opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: 
::: 00101: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=53 bytes
:::   --> EFI variable SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c: 1 bytes of data
:::   sha256     115aa827dbccfb44d216ad9ecfda56bdea620b860a94bed5b7a27bba1c4d02d8
:::   Data:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 0a 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 a.............+.................
:::         0020  53 00 65 00 63 00 75 00 72 00 65 00 42 00 6f 00 6f 00 74 00 00                                  S.e.c.u.r.e.B.o.o.t..
:::   Firmware hashed entire event data
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
:::   Remarshaled event for EFI variable SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 0a 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 a.............+.................
:::         0020  53 00 65 00 63 00 75 00 72 00 65 00 42 00 6f 00 6f 00 74 00 00                                  S.e.c.u.r.e.B.o.o.t..
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: 115aa827dbccfb44d216ad9ecfda56bdea620b860a94bed5b7a27bba1c4d02d8
::: 
::: 00168: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=36 bytes
:::   --> EFI variable PK-8be4df61-93ca-11d2-aa0d-00e098032b8c: 0 bytes of data
:::   sha256     dea7b80ab53a3daaa24d5cc46c64e1fa9ffd03739f90aadbd8c0867c4a5b4890
:::   Data:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  50 00 4b 00                                                                                     P.K.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "PK-8be4df61-93ca-11d2-aa0d-00e098032b8c"
:::   Remarshaled event for EFI variable PK-8be4df61-93ca-11d2-aa0d-00e098032b8c:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  50 00 4b 00                                                                                     P.K.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: dea7b80ab53a3daaa24d5cc46c64e1fa9ffd03739f90aadbd8c0867c4a5b4890
::: 
::: 001be: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=38 bytes
:::   --> EFI variable KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c: 0 bytes of data
:::   sha256     e670e121fcebd473b8bc41bb801301fc1d9afa33904f06f7149b74f12c47a68f
:::   Data:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  4b 00 45 00 4b 00                                                                               K.E.K.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c"
:::   Remarshaled event for EFI variable KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  4b 00 45 00 4b 00                                                                               K.E.K.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: e670e121fcebd473b8bc41bb801301fc1d9afa33904f06f7149b74f12c47a68f
::: 
::: 00216: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=36 bytes
:::   --> EFI variable db-d719b2cb-3d3a-4596-a3bc-dad00e67656f: 0 bytes of data
:::   sha256     baf89a3ccace52750c5f0128351e0422a41597a1adfd50822aa363b9d124ea7c
:::   Data:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00                                                                                     d.b.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "db-d719b2cb-3d3a-4596-a3bc-dad00e67656f"
:::   Remarshaled event for EFI variable db-d719b2cb-3d3a-4596-a3bc-dad00e67656f:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00                                                                                     d.b.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: baf89a3ccace52750c5f0128351e0422a41597a1adfd50822aa363b9d124ea7c
::: 
::: 0026c: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=38 bytes
:::   --> EFI variable dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f: 0 bytes of data
:::   sha256     9f75b6823bff6af1024a4e2036719cdd548d3cbc2bf1de8e7ef4d0ed01f94bf9
:::   Data:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00 78 00                                                                               d.b.x.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f"
:::   Remarshaled event for EFI variable dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00 78 00                                                                               d.b.x.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: 9f75b6823bff6af1024a4e2036719cdd548d3cbc2bf1de8e7ef4d0ed01f94bf9
::: 
::: 002c4: event type=EVENT_SEPARATOR pcr=7 digests=1 data=4 bytes
:::   sha256     df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119
:::   Data:
:::         0000  00 00 00 00                                                                                     ....
::: Extend PCR#7: sha256: df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119
::: 
::: 00c45: event type=EFI_VARIABLE_AUTHORITY pcr=7 digests=1 data=68 bytes
:::   --> EFI variable SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23: 18 bytes of data
:::   sha256     922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d
:::   Data:
:::         0000  50 ab 5d 60 46 e0 00 43 ab b6 3d d8 10 dd 8b 23 09 00 00 00 00 00 00 00 12 00 00 00 00 00 00 00 P.]`F..C..=....#................
:::         0020  53 00 62 00 61 00 74 00 4c 00 65 00 76 00 65 00 6c 00 73 62 61 74 2c 31 2c 32 30 32 31 30 33 30 S.b.a.t.L.e.v.e.l.sbat,1,2021030
:::         0040  32 31 38 0a                                                                                     218.
:::   Firmware hashed entire event data
::: Read 22 bytes from /sys/firmware/efi/efivars/SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23
:::   Remarshaled event for EFI variable SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23:
:::         0000  50 ab 5d 60 46 e0 00 43 ab b6 3d d8 10 dd 8b 23 09 00 00 00 00 00 00 00 12 00 00 00 00 00 00 00 P.]`F..C..=....#................
:::         0020  53 00 62 00 61 00 74 00 4c 00 65 00 76 00 65 00 6c 00 73 62 61 74 2c 31 2c 32 30 32 31 30 33 30 S.b.a.t.L.e.v.e.l.sbat,1,2021030
:::         0040  32 31 38 0a                                                                                     218.
::: Digest for EFI_VARIABLE_AUTHORITY did not change
::: Extend PCR#7: sha256: 922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d
sha256:7 b926225ac488e9c50ef2fa815aa7104b385a06907093bfb1dc62eeb7abecddf1
```

```
# cat /sys/class/tpm/tpm0/pcr-sha256/7
B926225AC488E9C50EF2FA815AA7104B385A06907093BFB1DC62EEB7ABECDDF1
```

```
- EventNum: 4
  PCRIndex: 7
  EventType: EV_EFI_VARIABLE_DRIVER_CONFIG
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "115aa827dbccfb44d216ad9ecfda56bdea620b860a94bed5b7a27bba1c4d02d8"
  EventSize: 53
  Event:
    VariableName: 8be4df61-93ca-11d2-aa0d-00e098032b8c
    UnicodeNameLength: 10
    VariableDataLength: 1
    UnicodeName: SecureBoot
    VariableData: "00"
- EventNum: 5
  PCRIndex: 7
  EventType: EV_EFI_VARIABLE_DRIVER_CONFIG
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "dea7b80ab53a3daaa24d5cc46c64e1fa9ffd03739f90aadbd8c0867c4a5b4890"
  EventSize: 36
  Event:
    VariableName: 8be4df61-93ca-11d2-aa0d-00e098032b8c
    UnicodeNameLength: 2
    VariableDataLength: 0
    UnicodeName: PK
- EventNum: 6
  PCRIndex: 7
  EventType: EV_EFI_VARIABLE_DRIVER_CONFIG
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "e670e121fcebd473b8bc41bb801301fc1d9afa33904f06f7149b74f12c47a68f"
  EventSize: 38
  Event:
    VariableName: 8be4df61-93ca-11d2-aa0d-00e098032b8c
    UnicodeNameLength: 3
    VariableDataLength: 0
    UnicodeName: KEK
- EventNum: 7
  PCRIndex: 7
  EventType: EV_EFI_VARIABLE_DRIVER_CONFIG
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "baf89a3ccace52750c5f0128351e0422a41597a1adfd50822aa363b9d124ea7c"
  EventSize: 36
  Event:
    VariableName: d719b2cb-3d3a-4596-a3bc-dad00e67656f
    UnicodeNameLength: 2
    VariableDataLength: 0
    UnicodeName: db
- EventNum: 8
  PCRIndex: 7
  EventType: EV_EFI_VARIABLE_DRIVER_CONFIG
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "9f75b6823bff6af1024a4e2036719cdd548d3cbc2bf1de8e7ef4d0ed01f94bf9"
  EventSize: 38
  Event:
    VariableName: d719b2cb-3d3a-4596-a3bc-dad00e67656f
    UnicodeNameLength: 3
    VariableDataLength: 0
    UnicodeName: dbx
- EventNum: 9
  PCRIndex: 7
  EventType: EV_SEPARATOR
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119"
- EventNum: 31
  PCRIndex: 7
  EventType: EV_EFI_VARIABLE_AUTHORITY
  DigestCount: 1
  Digests:
  - AlgorithmId: sha256
    Digest: "922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d"
  EventSize: 68
  Event:
    VariableName: 605dab50-e046-4300-abb6-3dd810dd8b23
    UnicodeNameLength: 9
    VariableDataLength: 18
    UnicodeName: SbatLevel
    VariableData: "736261742c312c323032313033303231380a"
```

aplanas commented:
Seems to be happening when several PCRs are tracked. For example, if I track {0,2,4} or {0,7} or {2,7}, or {7} it works, but {4,7} does not work.

```
 # pcr-oracle -d -d --from eventlog --tpm-eventlog /sys/kernel/security/tpm0/binary_bios_measurements predict 4,7
::: Initializing predictor for sha256:4,7 from eventlog
::: Detected TPMv2 event log
::: Successfully read 40 events from TPM event log
::: Created new predictor
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: runtime_read_efi_application(/dev/vda2, /EFI/systemd/shim.efi)
::: Read 965528 bytes from /tmp/efimnt.IabHx2//EFI/systemd/shim.efi
::: Reading EFI application (/dev/vda2)/EFI/systemd/shim.efi
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: runtime_read_efi_application(/dev/vda2, /EFI/systemd/grub.efi)
::: Read 125808 bytes from /tmp/efimnt.ayxuHY//EFI/systemd/grub.efi
::: Reading EFI application (/dev/vda2)/EFI/systemd/grub.efi
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: runtime_read_efi_application(/dev/vda2, /opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6)
::: Read 15485296 bytes from /tmp/efimnt.xFXD6t//opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6
::: Reading EFI application (/dev/vda2)/opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: 
::: 00101: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=53 bytes
:::   --> EFI variable SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c: 1 bytes of data
:::   sha256     115aa827dbccfb44d216ad9ecfda56bdea620b860a94bed5b7a27bba1c4d02d8
:::   Data:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 0a 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 a.............+.................
:::         0020  53 00 65 00 63 00 75 00 72 00 65 00 42 00 6f 00 6f 00 74 00 00                                  S.e.c.u.r.e.B.o.o.t..
:::   Firmware hashed entire event data
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
:::   Remarshaled event for EFI variable SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 0a 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 a.............+.................
:::         0020  53 00 65 00 63 00 75 00 72 00 65 00 42 00 6f 00 6f 00 74 00 00                                  S.e.c.u.r.e.B.o.o.t..
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: 115aa827dbccfb44d216ad9ecfda56bdea620b860a94bed5b7a27bba1c4d02d8
::: 
::: 00168: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=36 bytes
:::   --> EFI variable PK-8be4df61-93ca-11d2-aa0d-00e098032b8c: 0 bytes of data
:::   sha256     dea7b80ab53a3daaa24d5cc46c64e1fa9ffd03739f90aadbd8c0867c4a5b4890
:::   Data:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  50 00 4b 00                                                                                     P.K.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "PK-8be4df61-93ca-11d2-aa0d-00e098032b8c"
:::   Remarshaled event for EFI variable PK-8be4df61-93ca-11d2-aa0d-00e098032b8c:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  50 00 4b 00                                                                                     P.K.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: dea7b80ab53a3daaa24d5cc46c64e1fa9ffd03739f90aadbd8c0867c4a5b4890
::: 
::: 001be: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=38 bytes
:::   --> EFI variable KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c: 0 bytes of data
:::   sha256     e670e121fcebd473b8bc41bb801301fc1d9afa33904f06f7149b74f12c47a68f
:::   Data:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  4b 00 45 00 4b 00                                                                               K.E.K.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c"
:::   Remarshaled event for EFI variable KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  4b 00 45 00 4b 00                                                                               K.E.K.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: e670e121fcebd473b8bc41bb801301fc1d9afa33904f06f7149b74f12c47a68f
::: 
::: 00216: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=36 bytes
:::   --> EFI variable db-d719b2cb-3d3a-4596-a3bc-dad00e67656f: 0 bytes of data
:::   sha256     baf89a3ccace52750c5f0128351e0422a41597a1adfd50822aa363b9d124ea7c
:::   Data:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00                                                                                     d.b.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "db-d719b2cb-3d3a-4596-a3bc-dad00e67656f"
:::   Remarshaled event for EFI variable db-d719b2cb-3d3a-4596-a3bc-dad00e67656f:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00                                                                                     d.b.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: baf89a3ccace52750c5f0128351e0422a41597a1adfd50822aa363b9d124ea7c
::: 
::: 0026c: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=38 bytes
:::   --> EFI variable dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f: 0 bytes of data
:::   sha256     9f75b6823bff6af1024a4e2036719cdd548d3cbc2bf1de8e7ef4d0ed01f94bf9
:::   Data:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00 78 00                                                                               d.b.x.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f"
:::   Remarshaled event for EFI variable dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00 78 00                                                                               d.b.x.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: 9f75b6823bff6af1024a4e2036719cdd548d3cbc2bf1de8e7ef4d0ed01f94bf9
::: 
::: 002c4: event type=EVENT_SEPARATOR pcr=7 digests=1 data=4 bytes
:::   sha256     df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119
:::   Data:
:::         0000  00 00 00 00                                                                                     ....
::: Extend PCR#7: sha256: df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119
::: 
::: 0071e: event type=EFI_ACTION pcr=4 digests=1 data=40 bytes
:::   sha256     3d6772b4f84ed47595d72a2c4c5ffd15f5bb72c7507fe26f2aaee2c69d5633ba
:::   Data:
:::         0000  43 61 6c 6c 69 6e 67 20 45 46 49 20 41 70 70 6c 69 63 61 74 69 6f 6e 20 66 72 6f 6d 20 42 6f 6f Calling.EFI.Application.from.Boo
:::         0020  74 20 4f 70 74 69 6f 6e                                                                         t.Option
::: Extend PCR#4: sha256: 3d6772b4f84ed47595d72a2c4c5ffd15f5bb72c7507fe26f2aaee2c69d5633ba
::: 
::: 00850: event type=EVENT_SEPARATOR pcr=4 digests=1 data=4 bytes
:::   sha256     df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119
:::   Data:
:::         0000  00 00 00 00                                                                                     ....
::: Extend PCR#4: sha256: df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119
::: 
::: 00b08: event type=EFI_BOOT_SERVICES_APPLICATION pcr=4 digests=1 data=150 bytes
::: Boot Service Application; device path:
:::   ACPI       PNP0A03
:::   PCI        02.3
:::   PCI        00.0
:::   harddisk   part-uuid=d4c27816-f30b-4c47-9f39-b92e548307bc
:::   file-path  "/EFI/systemd/shim.efi"
:::   end
:::   sha256     211669e51a5e8c2315afe7a978740a972d721116ab81cbe384f993301ecde884
:::   Data:
:::         0000  18 80 fa 7d 00 00 00 00 98 bb 0e 00 00 00 00 00 00 00 00 00 00 00 00 00 76 00 00 00 00 00 00 00 ...}....................v.......
:::         0020  02 01 0c 00 d0 41 03 0a 00 00 00 00 01 01 06 00 03 02 01 01 06 00 00 00 04 01 2a 00 02 00 00 00 .....A....................*.....
:::         0040  00 18 00 00 00 00 00 00 00 a0 0f 00 00 00 00 00 16 78 c2 d4 0b f3 47 4c 9f 39 b9 2e 54 83 07 bc .................x....GL.9..T...
:::         0060  02 02 04 04 30 00 5c 00 45 00 46 00 49 00 5c 00 73 00 79 00 73 00 74 00 65 00 6d 00 64 00 5c 00 ....0.\.E.F.I.\.s.y.s.t.e.m.d.\.
:::         0080  73 00 68 00 69 00 6d 00 2e 00 65 00 66 00 69 00 00 00 7f ff 04 00                               s.h.i.m...e.f.i.......
::: Inspecting EFI application /dev/vda2(/EFI/systemd/grub.efi)
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: Computing authenticode digest using built-in PECOFF parser
::: Digest for EFI Boot Service Application /EFI/systemd/shim.efi did not change
::: Extend PCR#4: sha256: 211669e51a5e8c2315afe7a978740a972d721116ab81cbe384f993301ecde884
::: 
::: 00c45: event type=EFI_VARIABLE_AUTHORITY pcr=7 digests=1 data=68 bytes
:::   --> EFI variable SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23: 18 bytes of data
:::   sha256     922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d
:::   Data:
:::         0000  50 ab 5d 60 46 e0 00 43 ab b6 3d d8 10 dd 8b 23 09 00 00 00 00 00 00 00 12 00 00 00 00 00 00 00 P.]`F..C..=....#................
:::         0020  53 00 62 00 61 00 74 00 4c 00 65 00 76 00 65 00 6c 00 73 62 61 74 2c 31 2c 32 30 32 31 30 33 30 S.b.a.t.L.e.v.e.l.sbat,1,2021030
:::         0040  32 31 38 0a                                                                                     218.
:::   Firmware hashed entire event data
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: Unable to read EFI variable "SbatPolicy-605dab50-e046-4300-abb6-3dd810dd8b23"
::: Read 22 bytes from /sys/firmware/efi/efivars/SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23
::: Current SBAT datestampe: 2021030218
::: Candidate SBAT datestampe: 2023012900
::: Use candidate SbatLevel
:::   Remarshaled event for EFI variable SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23:
:::         0000  50 ab 5d 60 46 e0 00 43 ab b6 3d d8 10 dd 8b 23 09 00 00 00 00 00 00 00 2e 00 00 00 00 00 00 00 P.]`F..C..=....#................
:::         0020  53 00 62 00 61 00 74 00 4c 00 65 00 76 00 65 00 6c 00 73 62 61 74 2c 31 2c 32 30 32 33 30 31 32 S.b.a.t.L.e.v.e.l.sbat,1,2023012
:::         0040  39 30 30 0a 73 68 69 6d 2c 32 0a 67 72 75 62 2c 33 0a 67 72 75 62 2e 64 65 62 69 61 6e 2c 34 0a 900.shim,2.grub,3.grub.debian,4.
::: Digest for EFI_VARIABLE_AUTHORITY changed
:::   Old digest: sha256: 922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d
:::   New digest: sha256: e8e9578f5951ef16b1c1aa18ef02944b8375ec45ed4b5d8cdb30428db4a31016
::: Extend PCR#7: sha256: e8e9578f5951ef16b1c1aa18ef02944b8375ec45ed4b5d8cdb30428db4a31016
::: 
::: 00cfc: event type=EFI_BOOT_SERVICES_APPLICATION pcr=4 digests=1 data=84 bytes
::: Boot Service Application; device path:
:::   file-path  "/EFI/systemd/grub.efi"
:::   end
:::   sha256     4960527cbc306d6df965f443979949d121df2c856833a0fb0b37b95369cc9351
:::   Data:
:::         0000  18 50 07 7e 00 00 00 00 70 eb 01 00 00 00 00 00 00 00 30 01 01 00 00 00 34 00 00 00 00 00 00 00 .P.~....p.........0.....4.......
:::         0020  04 04 30 00 5c 00 45 00 46 00 49 00 5c 00 73 00 79 00 73 00 74 00 65 00 6d 00 64 00 5c 00 67 00 ..0.\.E.F.I.\.s.y.s.t.e.m.d.\.g.
:::         0040  72 00 75 00 62 00 2e 00 65 00 66 00 69 00 00 00 7f ff 04 00                                     r.u.b...e.f.i.......
::: Inspecting EFI application /dev/vda2(/opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6)
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: Computing authenticode digest using built-in PECOFF parser
::: Digest for EFI Boot Service Application /EFI/systemd/grub.efi did not change
::: Extend PCR#4: sha256: 4960527cbc306d6df965f443979949d121df2c856833a0fb0b37b95369cc9351
::: 
::: 00dd4: event type=EFI_BOOT_SERVICES_APPLICATION pcr=4 digests=1 data=276 bytes
::: Boot Service Application; device path:
:::   ACPI       PNP0A03
:::   PCI        02.3
:::   PCI        00.0
:::   harddisk   part-uuid=d4c27816-f30b-4c47-9f39-b92e548307bc
:::   file-path  "/opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6"
:::   end
:::   sha256     ca0f159b072f10b19dd41b6c64af9686d480691976921beac93a49c01bc014cb
:::   Data:
:::         0000  18 00 13 7b 00 00 00 00 70 49 ec 00 00 00 00 00 00 00 00 00 00 00 00 00 f4 00 00 00 00 00 00 00 ...{....pI......................
:::         0020  02 01 0c 00 d0 41 03 0a 00 00 00 00 01 01 06 00 03 02 01 01 06 00 00 00 04 01 2a 00 02 00 00 00 .....A....................*.....
:::         0040  00 18 00 00 00 00 00 00 00 a0 0f 00 00 00 00 00 16 78 c2 d4 0b f3 47 4c 9f 39 b9 2e 54 83 07 bc .................x....GL.9..T...
:::         0060  02 02 04 04 ae 00 5c 00 6f 00 70 00 65 00 6e 00 73 00 75 00 73 00 65 00 2d 00 74 00 75 00 6d 00 ......\.o.p.e.n.s.u.s.e.-.t.u.m.
:::         0080  62 00 6c 00 65 00 77 00 65 00 65 00 64 00 5c 00 36 00 2e 00 31 00 33 00 2e 00 35 00 2d 00 31 00 b.l.e.w.e.e.d.\.6...1.3...5.-.1.
:::         00a0  2d 00 64 00 65 00 66 00 61 00 75 00 6c 00 74 00 5c 00 6c 00 69 00 6e 00 75 00 78 00 2d 00 37 00 -.d.e.f.a.u.l.t.\.l.i.n.u.x.-.7.
:::         00c0  35 00 62 00 34 00 65 00 32 00 61 00 32 00 39 00 35 00 62 00 31 00 31 00 33 00 30 00 38 00 64 00 5.b.4.e.2.a.2.9.5.b.1.1.3.0.8.d.
:::         00e0  30 00 38 00 65 00 37 00 37 00 65 00 62 00 34 00 35 00 30 00 37 00 38 00 30 00 31 00 30 00 63 00 0.8.e.7.7.e.b.4.5.0.7.8.0.1.0.c.
:::         0100  38 00 32 00 34 00 36 00 64 00 66 00 36 00 00 00 7f ff 04 00                                     8.2.4.6.d.f.6.......
::: Computing authenticode digest using built-in PECOFF parser
::: Digest for EFI Boot Service Application /opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6 did not change
::: Extend PCR#4: sha256: ca0f159b072f10b19dd41b6c64af9686d480691976921beac93a49c01bc014cb
sha256:4 ff923809ad784bf6de4b76128eb856854dd33c07591e3e5e5be7c2561ae86893
sha256:7 ae924efd4be5f1aaa823e516560800e1239542753f22bdf632b2edfe7c5b13d3
```

If I read it properly, the issue is happening here:

```
::: 00c45: event type=EFI_VARIABLE_AUTHORITY pcr=7 digests=1 data=68 bytes
:::   --> EFI variable SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23: 18 bytes of data
:::   sha256     922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d
:::   Data:
:::         0000  50 ab 5d 60 46 e0 00 43 ab b6 3d d8 10 dd 8b 23 09 00 00 00 00 00 00 00 12 00 00 00 00 00 00 00 P.]`F..C..=....#................
:::         0020  53 00 62 00 61 00 74 00 4c 00 65 00 76 00 65 00 6c 00 73 62 61 74 2c 31 2c 32 30 32 31 30 33 30 S.b.a.t.L.e.v.e.l.sbat,1,2021030
:::         0040  32 31 38 0a                                                                                     218.
:::   Firmware hashed entire event data
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: Unable to read EFI variable "SbatPolicy-605dab50-e046-4300-abb6-3dd810dd8b23"
::: Read 22 bytes from /sys/firmware/efi/efivars/SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23
::: Current SBAT datestampe: 2021030218
::: Candidate SBAT datestampe: 2023012900
::: Use candidate SbatLevel
:::   Remarshaled event for EFI variable SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23:
:::         0000  50 ab 5d 60 46 e0 00 43 ab b6 3d d8 10 dd 8b 23 09 00 00 00 00 00 00 00 2e 00 00 00 00 00 00 00 P.]`F..C..=....#................
:::         0020  53 00 62 00 61 00 74 00 4c 00 65 00 76 00 65 00 6c 00 73 62 61 74 2c 31 2c 32 30 32 33 30 31 32 S.b.a.t.L.e.v.e.l.sbat,1,2023012
:::         0040  39 30 30 0a 73 68 69 6d 2c 32 0a 67 72 75 62 2c 33 0a 67 72 75 62 2e 64 65 62 69 61 6e 2c 34 0a 900.shim,2.grub,3.grub.debian,4.
::: Digest for EFI_VARIABLE_AUTHORITY changed
:::   Old digest: sha256: 922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d
:::   New digest: sha256: e8e9578f5951ef16b1c1aa18ef02944b8375ec45ed4b5d8cdb30428db4a31016
::: Extend PCR#7: sha256: e8e9578f5951ef16b1c1aa18ef02944b8375ec45ed4b5d8cdb30428db4a31016
```

The correct value is the current one: `922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d`, not the calculated one.  This miscalculation only appears when PCR 4 is included.

```
# pcr-oracle -d -d --from eventlog --tpm-eventlog /sys/kernel/security/tpm0/binary_bios_measurements predict 7
::: Initializing predictor for sha256:7 from eventlog
::: Detected TPMv2 event log
::: Successfully read 40 events from TPM event log
::: Created new predictor
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: runtime_read_efi_application(/dev/vda2, /EFI/systemd/shim.efi)
::: Read 965528 bytes from /tmp/efimnt.UNyyLr//EFI/systemd/shim.efi
::: Reading EFI application (/dev/vda2)/EFI/systemd/shim.efi
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: runtime_read_efi_application(/dev/vda2, /EFI/systemd/grub.efi)
::: Read 125808 bytes from /tmp/efimnt.MSPprG//EFI/systemd/grub.efi
::: Reading EFI application (/dev/vda2)/EFI/systemd/grub.efi
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: runtime_read_efi_application(/dev/vda2, /opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6)
::: Read 15485296 bytes from /tmp/efimnt.TC66XC//opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6
::: Reading EFI application (/dev/vda2)/opensuse-tumbleweed/6.13.5-1-default/linux-75b4e2a295b11308d08e77eb45078010c8246df6
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
::: 
::: 00101: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=53 bytes
:::   --> EFI variable SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c: 1 bytes of data
:::   sha256     115aa827dbccfb44d216ad9ecfda56bdea620b860a94bed5b7a27bba1c4d02d8
:::   Data:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 0a 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 a.............+.................
:::         0020  53 00 65 00 63 00 75 00 72 00 65 00 42 00 6f 00 6f 00 74 00 00                                  S.e.c.u.r.e.B.o.o.t..
:::   Firmware hashed entire event data
::: Read 5 bytes from /sys/firmware/efi/efivars/SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c
:::   Remarshaled event for EFI variable SecureBoot-8be4df61-93ca-11d2-aa0d-00e098032b8c:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 0a 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 a.............+.................
:::         0020  53 00 65 00 63 00 75 00 72 00 65 00 42 00 6f 00 6f 00 74 00 00                                  S.e.c.u.r.e.B.o.o.t..
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: 115aa827dbccfb44d216ad9ecfda56bdea620b860a94bed5b7a27bba1c4d02d8
::: 
::: 00168: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=36 bytes
:::   --> EFI variable PK-8be4df61-93ca-11d2-aa0d-00e098032b8c: 0 bytes of data
:::   sha256     dea7b80ab53a3daaa24d5cc46c64e1fa9ffd03739f90aadbd8c0867c4a5b4890
:::   Data:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  50 00 4b 00                                                                                     P.K.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "PK-8be4df61-93ca-11d2-aa0d-00e098032b8c"
:::   Remarshaled event for EFI variable PK-8be4df61-93ca-11d2-aa0d-00e098032b8c:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  50 00 4b 00                                                                                     P.K.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: dea7b80ab53a3daaa24d5cc46c64e1fa9ffd03739f90aadbd8c0867c4a5b4890
::: 
::: 001be: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=38 bytes
:::   --> EFI variable KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c: 0 bytes of data
:::   sha256     e670e121fcebd473b8bc41bb801301fc1d9afa33904f06f7149b74f12c47a68f
:::   Data:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  4b 00 45 00 4b 00                                                                               K.E.K.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c"
:::   Remarshaled event for EFI variable KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c:
:::         0000  61 df e4 8b ca 93 d2 11 aa 0d 00 e0 98 03 2b 8c 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 a.............+.................
:::         0020  4b 00 45 00 4b 00                                                                               K.E.K.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: e670e121fcebd473b8bc41bb801301fc1d9afa33904f06f7149b74f12c47a68f
::: 
::: 00216: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=36 bytes
:::   --> EFI variable db-d719b2cb-3d3a-4596-a3bc-dad00e67656f: 0 bytes of data
:::   sha256     baf89a3ccace52750c5f0128351e0422a41597a1adfd50822aa363b9d124ea7c
:::   Data:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00                                                                                     d.b.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "db-d719b2cb-3d3a-4596-a3bc-dad00e67656f"
:::   Remarshaled event for EFI variable db-d719b2cb-3d3a-4596-a3bc-dad00e67656f:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00                                                                                     d.b.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: baf89a3ccace52750c5f0128351e0422a41597a1adfd50822aa363b9d124ea7c
::: 
::: 0026c: event type=EFI_VARIABLE_DRIVER_CONFIG pcr=7 digests=1 data=38 bytes
:::   --> EFI variable dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f: 0 bytes of data
:::   sha256     9f75b6823bff6af1024a4e2036719cdd548d3cbc2bf1de8e7ef4d0ed01f94bf9
:::   Data:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00 78 00                                                                               d.b.x.
:::   Firmware hashed entire event data
::: Unable to read EFI variable "dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f"
:::   Remarshaled event for EFI variable dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f:
:::         0000  cb b2 19 d7 3a 3d 96 45 a3 bc da d0 0e 67 65 6f 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ....:=.E.....geo................
:::         0020  64 00 62 00 78 00                                                                               d.b.x.
::: Digest for EFI_VARIABLE_DRIVER_CONFIG did not change
::: Extend PCR#7: sha256: 9f75b6823bff6af1024a4e2036719cdd548d3cbc2bf1de8e7ef4d0ed01f94bf9
::: 
::: 002c4: event type=EVENT_SEPARATOR pcr=7 digests=1 data=4 bytes
:::   sha256     df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119
:::   Data:
:::         0000  00 00 00 00                                                                                     ....
::: Extend PCR#7: sha256: df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119
::: 
::: 00c45: event type=EFI_VARIABLE_AUTHORITY pcr=7 digests=1 data=68 bytes
:::   --> EFI variable SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23: 18 bytes of data
:::   sha256     922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d
:::   Data:
:::         0000  50 ab 5d 60 46 e0 00 43 ab b6 3d d8 10 dd 8b 23 09 00 00 00 00 00 00 00 12 00 00 00 00 00 00 00 P.]`F..C..=....#................
:::         0020  53 00 62 00 61 00 74 00 4c 00 65 00 76 00 65 00 6c 00 73 62 61 74 2c 31 2c 32 30 32 31 30 33 30 S.b.a.t.L.e.v.e.l.sbat,1,2021030
:::         0040  32 31 38 0a                                                                                     218.
:::   Firmware hashed entire event data
::: Read 22 bytes from /sys/firmware/efi/efivars/SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23
:::   Remarshaled event for EFI variable SbatLevelRT-605dab50-e046-4300-abb6-3dd810dd8b23:
:::         0000  50 ab 5d 60 46 e0 00 43 ab b6 3d d8 10 dd 8b 23 09 00 00 00 00 00 00 00 12 00 00 00 00 00 00 00 P.]`F..C..=....#................
:::         0020  53 00 62 00 61 00 74 00 4c 00 65 00 76 00 65 00 6c 00 73 62 61 74 2c 31 2c 32 30 32 31 30 33 30 S.b.a.t.L.e.v.e.l.sbat,1,2021030
:::         0040  32 31 38 0a                                                                                     218.
::: Digest for EFI_VARIABLE_AUTHORITY did not change
::: Extend PCR#7: sha256: 922e939a5565798a5ef12fe09d8b49bf951a8e7f89a0cca7a51636693d41a34d
sha256:7 b926225ac488e9c50ef2fa815aa7104b385a06907093bfb1dc62eeb7abecddf1
```

lcp commented:
Ah, when encountering SbatLevelRT, my patch relies on the PCR4 shim event to extract the .sbatlevel section inside shim.efi, so it introduces PCR4 dependency implicitly...

I checked my patch again and found the sbatlevel candidate is not correct when Secure boot is disabled. Will fix it soon.

lcp committed:
Fix SbatLevelRT prediction when Secure Boot is disabled

Since shim 15.8, instead of using SBAT automatic as the default
candidate, it always resets SbatLevel to SBAT_ORIGINAL. To make the
prediction work for shim >= 15.8, the additinal check is added to adjust
the SBAT candidate.

Also fix POLICY_RESET for not setting sbat_reset to true and a few
typos.

Signed-off-by: Gary Lin <glin@suse.com>

lcp commented:
Hmmm, shim 15.8 seems a bit unpredictable when Secure Boot is disabled.

Since shim 15.8, it always resets SbatLevel to SBAT_ORIGINAL(`"sbat,1,2021030218\n"`) if Secure Boot is disabled and no SbatPolicy is specified. If SbatPolicy is set, shim follows the variable to choose the SBAT candidate. However, SbatPolicy will be deleted immediately after shim reads it. If pcr-oracle does the prediction of SbatLevelRT when SbatPolicy is set, the prediction works for the next boot. But, for the next of the next, the prediction may not work anymore...

aplanas commented:
uhm then I do not understand the issue ... in my tests the event log did not change, but the pcr-oracle prediction changed for sbatpolicy based if PCR4 was included or not

lcp commented:
> uhm then I do not understand the issue ... in my tests the event log did not change, but the pcr-oracle prediction changed for sbatpolicy based if PCR4 was included or not

I updated the PR to address the problem you had.

What I am worry about is SbatPolicy. The variable may be set by mokutil in %post script of shim and it's also the timing that the sealed key gets updated. When pcr-oracle does the prediction after shim update, it predicts SbatLevelRT based on the new shim.efi + SbatPolicy. But SbatPolicy will be cleared by shim during the next boot, and SbatLevel could be reset to SBAT_ORIGINAL for the later boots, and then PCR7 changes again. This could fail key unsealing unexpectedly :-(

lcp committed:
Stop the SbatLevelRT prediction if .sbatlevel not available

The SbatLevelRT prediction relies on the PCR4 events to locate shim.efi,
so PCR4 has to be a hard requirement.

Signed-off-by: Gary Lin <glin@suse.com>
